support type annotations on global variables

simeonschaub opened the pull request:
This is an initial proposal for supporting typed globals based on #43455.
The syntax for this is either `global x::T` or just doing `x::T = 1` in
global scope. It is even supported to add these annotations to globals
from inside functions. ~~The type declaration will then be applied when
the method is defined and inside the function body conversion to the
specified type will happen automatically, similar to type annotations
for local variables.~~

~~This conversion will not be applied if the assignment is not done inside
the same scope of the type annotation however. This could potentially be
supported as well, but the problem is that this would mean any
assignment to a global variable - typed or not - would need to go
through a call to `convert` first, since lowering can't know if a given
binding already has a type annotation. This probably wouldn't be a good
thing for latency and sysimage size and would add an invalidation risk.~~

Any assignment to a global now adds a conversion step and a type
assertion, similar to how type annotations work inside local scopes.

It is allowed to refine a type annotation, but there will be a warning.
Widening the type or changing it to any other type that is not a subtype
of the previous one is an error, since that could cause cached compiled
code to become invalid.

replaces #43455
closes #964

miguelraz committed:
add typing to global bindings

new types should go at end of struct

simeonschaub committed:
support type annotations on global variables

This is an initial proposal for supporting typed globals based on #43455.
The syntax for this is either `global x::T` or just doing `x::T = 1` in
global scope. It is even supported to add these annotations to globals
from inside functions. The type declaration will then be applied when
the method is defined and inside the function body conversion to the
specified type will happen automatically, similar to type annotations
for local variables.

This conversion will not be applied if the assignment is not done inside
the same scope of the type annotation however. This could potentially be
supported as well, but the problem is that this would mean any
assignment to a global variable - typed or not - would need to go
through a call to `convert` first, since lowering can't know if a given
binding already has a type annotation. This probably wouldn't be a good
thing for latency and sysimage size and would add an invalidation risk.

It is allowed to refine a type annotation, but there will be a warning.
Widening the type or changing it to any other type that is not a subtype
of the previous one is an error, since that could cause cached compiled
code to become invalid.

replaces #43455
closes #964

JeffBezanson commented:
> This conversion will not be applied if the assignment is not done inside
> the same scope of the type annotation however. This could potentially be
> supported as well

I think we will need to support this. Users would expect to write `x::Int = 0` at the top level once, and have that apply everywhere. I know it's a bit annoying, which is why we haven't done this yet :smile: But assigning to globals should already be quite rare and slow, so the extra convert call is fine with me.

JeffBezanson commented on src/builtins.c:
```diff
@@ -1655,6 +1655,36 @@ JL_CALLABLE(jl_f__equiv_typedef)
     return equiv_type(args[0], args[1]) ? jl_true : jl_false;
 }
 
+JL_CALLABLE(jl_f__set_typeof)
```
```suggestion
JL_CALLABLE(jl_f__set_binding_type)
```

simeonschaub committed:
rename to `_set_binding_type!`

simeonschaub committed:
try to fix analyzegc failure

simeonschaub committed:
always add conversion step for globals

KristofferC commented:
> It is even supported to add these annotations to globals
from inside functions. The type declaration will then be applied when
the method is defined and inside the function body conversion to the
specified type will happen automatically, similar to type annotations
for local variables.

That seems like a quite weird feature to me. Was this implemented because it was easy or does it have some good use cases?

simeonschaub commented:
We could disallow adding these type annotations inside local scopes, but since we already allow defining new globals from local scopes, I think it's only consistent to allow declaring typed ones as well. This was a bit of a special case before my last commit, since lowering would then use that annotation to add the automatic conversion. Now, this automatic conversion is always added on assignments to globals though, so there's not really much special handling except that we explicitly tell lowering to always set binding types from the toplevel via the `toplevel-bufirst` expression head.

simeonschaub committed:
try to fix tests

simeonschaub committed:
fix inference test

Lowering split up `i = i - 1` using an additional SSA value here, which
should be completely harmless, but broke this test.

simeonschaub committed:
fix analyzegc failures

antoine-levitt commented:
Very cool, great feature! Is this limited to literal types, or will it be possible to do `T = Int32; x::T`? If so, we could introduce a `@typeconst` macro (or even language keyword?) and have `@typeconst x=2` lower to `temp = 2; x::typeof(temp) = temp`; that would get rid of the most common reason for the misuse of `const` (at least for me). (then, in a perfect world, it would be allowed to redefine the type of global variables, and finally have performant and flexible global variables...)

simeonschaub commented:
> Is this limited to literal types, or will it be possible to do `T = Int32; x::T`?

Any valid top level expression is fine.

I'm not sure we need such a `@typeconst` macro in base, but that can always be discussed in a future issue/PR.

simeonschaub commented:
/buildkite rerun failed

simeonschaub committed:
allow `Core._get_binding_type` to be elided

simeonschaub committed:
clean up `julia-syntax.scm`

simeonschaub commented:
Ok, I have now implemented Jeff's changed this to always add a conversion step for assignments to globals (except for when the global is also declared as `const`). I have also added some inference support for `Core._get_binding_type` so that assignment to typed globals should now generally be type stable.

One aspect I would appreciate some input on is thread safety. I think it should be fine to leave the `ty` field in `jl_binding_t` non-atomic since I'd consider redefining the binding type after it's been declared not well defined and only allowed in some cases for the sake of convenience in interactive use. I am a bit worried there could be potential for a race condition if a thread sets a binding type while a different one sets its value to one of an incompatible type. I wasn't able to trigger that in my limited testing, but that probably doesn't mean much.

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

simeonschaub committed:
add docs

simeonschaub committed:
add NEWS entry

simeonschaub committed:
remove leading underscores in names

simeonschaub commented:
Ok, I have now added some docs and a NEWS entry. Everything should be ready to review now.

simeonschaub committed:
fix docs

simeonschaub committed:
just remove ref to `<:`

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

simeonschaub commented:
It looks like some packages do dodgy things with `convert(::Type{Any}, x)`: https://juliahub.com/ui/Search?q=convert%28%3A%3AType%7BAny%7D&type=code. Probably a good idea to run a PkgEval run

JeffBezanson commented on src/builtins.c:
```diff
@@ -1655,6 +1655,46 @@ JL_CALLABLE(jl_f__equiv_typedef)
+    jl_value_t *ty = args[2];
+    JL_TYPECHK(set_binding_type!, type, ty);
+    jl_binding_t *b = jl_get_binding_wr((jl_module_t*)args[0], (jl_sym_t*)args[1], 1);
+    if (b->constp && ty != (jl_value_t*)jl_any_type) {
```
This check seems unnecessary. Sure, declaring a type on a constant is kind of redundant but we might as well allow it.

simeonschaub replied:
Yeah, I think we can remove that check. Should declaring typed globals as const also be allowed then?

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4015,30 +4039,40 @@ f(x) = yt(x)
+                  (cl-convert
+                    (let ((ref (binding-to-globalref (cadr e))))
+                      (and ref (get globals ref #f)
+                           (error (string "multiple type annotations for global \""
```
I think we can remove this error and just let it get caught by the run time check. There are a few reasons for this. (1) Global declarations work like top-level effects, e.g.
```
let
    global const x = 1
    if false
        global const x = 2
    end
end
```
does not error since the second declaration doesn't execute.
(2) It disallows things that are allowed at the top level, e.g.
```
let
    global y::Int = 1
    global y::Int = 2
end
```
errors, but without the `let` it's fine.

JeffBezanson replied:
Though come to think of it, we should probably disallow type declarations on globals inside functions, like we do for `const` declarations.

simeonschaub replied:
My idea was to mostly follow the behavior for type annotations in local scopes here, so that adding a `global x` inside a function should change as little as possible. I am not sure it's exactly the same as `const` since there is no local-scope equivalent of `const`. I am not super adamant about this though, we can be more lenient here or allow these in local scope altogether if you think that's more intuitive.

JeffBezanson replied:
I think the main driver is the kind of variable in question, global vs. local. Almost everything about them is different. So I think all declarations on globals (type or const) should work in a way that makes sense for globals, and all declarations on locals should make sense for them.

oscardssmith replied:
One reason to allow type declarations of globals inside functions is that sometimes, figuring out the type is difficult from a global scope without introducing a new variable. For a dumb example, `f = x->x+1` is very hard to type annotate from global scope without introducing a second global variable.

simeonschaub replied:
Ok, I will make this a runtime check.
@oscardssmith That usecase is not supported though, the type declaration always needs to be moved to the toplevel in lowering, otherwise inference can't access that type information.

vtjnash commented:
Jeff should review the julia-syntax.scm changes

simeonschaub commented:
@nanosoldier `runtests(ALL, vs = ":master")`

simeonschaub commented on base/essentials.jl:
```diff
@@ -161,6 +161,11 @@ end
 
 function iterate end
 
+if @isdefined(convert)
+else
```
The reason I did it that way was because `!` is not defined during bootstrap yet. I can try moving this to `compiler/compiler.jl` though.

simeonschaub commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1805,7 +1805,7 @@ function abstract_eval_global(M::Module, s::Symbol)
         return Const(getfield(M,s))
     end
-    return Any
+    return Core.get_binding_type(M, s)
```
No, this should never throw. Do you mean it's impure because the binding type can be redefined? That's the same issue we have with constants already.

vtjnash replied:
It is impure because it can start without a binding type, and be added later.

simeonschaub replied:
Yes, I agree that it's a little iffy. If we don't mark it as pure though, the `convert` step will always be dynamic. Like I said, my thinking was that this is similar to how we assume purity of `const` values when resolving `GlobalRef`s.

vtjnash replied:
yes, we need to model it more precisely, so that it knows whether the global has a type and/or value assigned yet

vtjnash replied:
potentially we can make it syntactic, so that you know, as soon as the method is defined, whether there will be a type already given or not on the global

vtjnash replied:
(which would make it pure)

simeonschaub replied:
> potentially we can make it syntactic, so that you know, as soon as the method is defined, whether there will be a type already given or not on the global

Could you elaborate on that? Do you mean we basically make a substitution right after lowering?

simeonschaub replied:
If I understand you correctly, wouldn't this approach fail in cases like:

```julia
begin
    global x::Int
    x = 1.0
end
```

Or maybe we could add back some logic I had here previously to analyze cases like these in the frontend? 🤔

vtjnash replied:
yes, doing inference on code like that is already buggy for that reason

We could make the substitution early, but I actually mean that a `global` decl should simultaneously set the type (to Any), except where the global already has a type. This would prohibit code ordering like:
```julia
f() = (global x; x += 1)
x::Int = 3
```

But I think it is probably better for the global type declarations to appear first (before their usage) anyways

simeonschaub replied:
Ok, that's interesting. You mean, we would fix it to `Any` anytime there's a write to a global? Or should `f() = (global x; x)` already count as fixing the type to `Any`?

vtjnash replied:
`f() = (global x; x)` already means the same as `global x; f() = x`, so both would fix the type to Any.

simeonschaub replied:
`x` is never written to, only read from in that example though. I thought we'd only fix that type if the function actually writes to the binding, since we only need the binding type if we actually do an assignment.

simeonschaub replied:
Ok, I have made that change now, but note that the type is only fixed if `x` is actually written to.

simeonschaub commented on src/builtins.c:
```diff
@@ -1655,6 +1655,46 @@ JL_CALLABLE(jl_f__equiv_typedef)
+    if (b->constp && ty != (jl_value_t*)jl_any_type) {
+        jl_errorf("cannot set type for constant %s", jl_symbol_name(b->name));
+    }
+    jl_value_t *val = jl_atomic_load_relaxed(&b->value);
+    if (val && !jl_isa(val, ty)) {
+        jl_errorf("cannot set type for global %s. It already has a value of a different type.",
+                  jl_symbol_name(b->name));
+    }
```
Ok, I can make that change. Does the atomic load here need to be changed as well then?

vtjnash replied:
we can remove the load of `value`, yes

and will need to make the `typ` field to be `_Atomic` annotated in julia.h

simeonschaub replied:
Ah, so you mean we never allow changing the binding type once `value` is defined?

simeonschaub replied:
I tried to implement this in https://github.com/JuliaLang/julia/pull/43671/commits/1d54432c40712f42b48f8903dd9fd21e705ea7bc, but somehow `get_binding_type` now seems to return invalid types. @vtjnash Any ideas what I did wrong here?

simeonschaub commented on src/builtins.c:
```diff
@@ -1655,6 +1655,46 @@ JL_CALLABLE(jl_f__equiv_typedef)
+    }
+    if (b->ty && ty != b->ty) {
+        if (jl_subtype(ty, b->ty)) {
+            jl_safe_printf("WARNING: redefinition of type of %s. This may fail, cause incorrect answers, or produce other errors.\n",
+                           jl_symbol_name(b->name));
```
Ok, I can change the warning to something less alarming. We could make it an error, but I am not sure that's necessary and I could still see this being useful in interactive use.

simeonschaub commented on src/julia.h:
```diff
@@ -511,6 +511,7 @@ typedef struct {
     uint8_t exportp:1;
     uint8_t imported:1;
     uint8_t deprecated:2; // 0=not deprecated, 1=renamed, 2=moved to another package
+    jl_value_t *ty;                  // add type for globals
```
Will do

miguelraz replied:
Wouldn't that break code that depends on the struct offsets? I know it's internal but just checking.

simeonschaub replied:
Technically yes, but this is a purely internal type and I don't think we generally make any strong guarantees about memory offsets of types.

miguelraz replied:
Let it rip then.

vtjnash commented on src/module.c:
```diff
@@ -375,6 +376,19 @@ JL_DLLEXPORT jl_value_t *jl_binding_owner(jl_module_t *m, jl_sym_t *var)
+    JL_LOCK(&m->lock);
+    jl_binding_t *b = (jl_binding_t*)ptrhash_get(&m->bindings, var);
+    if (b == HT_NOTFOUND || b->owner == NULL)
+        b = using_resolve_binding(m, var, NULL, 0);
```
how does this work for imported bindings, once we know the owner? I don't think I see any tests for nested modules and/or using/export/imports

simeonschaub replied:
Yes, I should add some tests. I believe this should already read the type of the correct binding, but I'll double check. Of course, there's still the same purity issue discussed above.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43671#issuecomment-1018824462) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/3656ddf_vs_816c6a2/report.html).

simeonschaub commented:
nanosoldier `runtests(["BioServices", "BlobTracking", "CorrelationFunctions", "CounterfactualRegret", "Cthulhu", "DICOM", "ExtendableGrids", "FaultDetectionTools", "FlashWeave", "GPLinearODEMaker", "GeoDatasets", "ITensors", "JuliaInterpreter", "Juniper", "LMDB", "LibCURL", "LogisticOptTools", "LoweredCodeUtils", "NTNk", "OpenDSSDirect", "OptimKit", "OteraEngine", "Plasma", "ProgressiveHedging", "SparseRegression", "Syslogs", "VertexSafeGraphs", "VoronoiGraph"], vs = ":master")`

simeonschaub commented:
@nanosoldier `runtests(["BioServices", "BlobTracking", "CorrelationFunctions", "CounterfactualRegret", "Cthulhu", "DICOM", "ExtendableGrids", "FaultDetectionTools", "FlashWeave", "GPLinearODEMaker", "GeoDatasets", "ITensors", "JuliaInterpreter", "Juniper", "LMDB", "LibCURL", "LogisticOptTools", "LoweredCodeUtils", "NTNk", "OpenDSSDirect", "OptimKit", "OteraEngine", "Plasma", "ProgressiveHedging", "SparseRegression", "Syslogs", "VertexSafeGraphs", "VoronoiGraph"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/43671#issuecomment-1019280837) has completed - possible new issues were detected. A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/3656ddf_vs_1db8b8f/report.html).

simeonschaub committed:
try to make `ty` field atomic

vtjnash commented:
you also likely need to add support in dump.c and staticdata.c for saving this field

simeonschaub commented:
Ok, will do. That doesn't seem to fix the segfaults though

simeonschaub committed:
(de)serialize type field correctly

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

simeonschaub committed:
fix serialization

simeonschaub committed:
address review comments

Change lowering s.t. types are always frozen before functions writing to
a global are defined. `get_binding_type` now errors if the binding is
still uninitialized. Also make multiple type declarations in local scope
a runtime error.

simeonschaub committed:
fix test, add test for different owner

simeonschaub committed:
try to fix distributed tests

simeonschaub committed:
fix whitespace snafu

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

simeonschaub committed:
refactor lowering a bit, fix unnecessary ssavalue

simeonschaub commented:
I believe I addressed all review comments now.

As @vtjnash suggested, the binding type is now frozen (to `Any` if no type is specified) as soon as either lowering detects the binding being written to or even if that's bypassed by calling `jl_checked_assignment` directly. Changing that type after the fact is now always an error. I have also changed `get_binding_type` s.t. it now errors if there is no type assigned to a given binding yet. This means that `get_binding_type` can now always be assumed to be pure if it doesn't throw.

The error when calling `set_binding_type!` on a constant and vice versa has also been removed and I added a test with bindings that are already imported.

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -3451,23 +3477,40 @@ f(x) = yt(x)
 ;; collect all toplevel-butfirst expressions inside `e`, and return
 ;; (ex . stmts), where `ex` is the expression to evaluated and
 ;; `stmts` is a list of statements to move to the top level.
+;; also expand set-binding-type expressions and handle them similarly
```
Looks like this still allows global type declarations in local scopes? Jameson and I agree we can just disallow that, which should simplify the code as well.

simeonschaub replied:
Ok, I have now disallowed that. Something like `f() = (global x = 1;)` still freezes the type of `x` to `Any` (unless already set) before `f` is ever executed though in order not to introduce dynamic calls to `convert`.

simeonschaub committed:
disallow type annotations in local scope

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

JeffBezanson commented on base/compiler/abstractinterpretation.jl:
```diff
@@ -1793,7 +1793,9 @@ function abstract_eval_global(M::Module, s::Symbol)
         return Const(getfield(M,s))
     end
-    return Any
+    ty = ccall(:jl_binding_type, Ptr{Cvoid}, (Any, Any), M, s)
```
This could return `nothing` instead of NULL to avoid `unsafe_pointer_to_objref`.

simeonschaub replied:
Ok, I'll change that

simeonschaub committed:
use nothing instead of NULL

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -3382,12 +3388,34 @@ f(x) = yt(x)
+                         ex
+                         `(toplevel-butfirst
+                            ,ex
+                            (call (core set_binding_type!) ,(cadr ref) (inert ,(caddr ref))))))
```
Could we call this in `resolve_globals` instead (where other `global x` declarations in functions are handled), so the front end doesn't need to handle it?

simeonschaub replied:
Hmm, yes, maybe. I don't know if we can distinguish writes from only reads from globals there though.

vtjnash replied:
`global x` is the same as a write, for syntactic purposes

simeonschaub replied:
Ok, I tried that, but the issue are toplevel expressions like `x = 1`. Those are writes to a global variable, but this is difficult to recognize in the interpreter before `get_binding_type` is evaluated, since there is no `global` expression.

JeffBezanson replied:
Maybe instead of calling get_binding_type, the front end could call get_binding_type_for_write, which sets the type to Any if there isn't one yet.

simeonschaub replied:
Yeah, that makes sense. I think we can just change he semantics of `get_binding_type` that way, since it should only be called through the frontend anyways.

simeonschaub replied:
Ok, what do you think of [a334a9b](https://github.com/JuliaLang/julia/pull/43671/commits/a334a9b06bcfe47e59d0d50dc82499d2f41f6461)?

simeonschaub committed:
address Jeff's review comment

simeonschaub committed:
fix doctests

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -4019,30 +4044,37 @@ f(x) = yt(x)
+                          (begin
+                            (put! globals ref #t)
+                            `(block
+                               (toplevel-only set_binding_type!)
```
This needs a case added to `head-to-text` to give a nicer error message.

simeonschaub replied:
Yes, good catch!

simeonschaub committed:
better error message in `head-to-text`

simeonschaub committed:
fix elision of `get_binding_type`

simeonschaub committed:
Merge remote-tracking branch 'origin/master' into sds/typedglobals

simeonschaub committed:
fix tests

simeonschaub committed:
update test for #33243

simeonschaub committed:
make test more robust

simeonschaub committed:
use binding type in codegen?

simeonschaub commented on src/codegen.cpp:
```diff
@@ -2385,7 +2385,7 @@ static jl_cgval_t emit_globalref(jl_codectx_t &ctx, jl_module_t *mod, jl_sym_t *
         v->setOrdering(AtomicOrdering::Unordered);
         tbaa_decorate(ctx.tbaa().tbaa_binding, v);
-        return mark_julia_type(ctx, v, true, (jl_value_t*)jl_any_type);
+        return mark_julia_type(ctx, v, true, bnd->ty);
```
I think this is the correct thing to do? Not very confident about this though...

JeffBezanson replied:
Yes this should work.

simeonschaub commented:
Was just about to press the button... 😂 Thanks for the thorough review and for bearing with me!

KristofferC commented:
Fixes a pre #1000 issue 👏
